pkg/controller/disruption/disruption.go

```diff
@@ -642,18 +656,38 @@ func (dc *DisruptionController) getExpectedScale(pdb *policy.PodDisruptionBudget
 	// A mapping from controllers to their scale.
 	controllerScale := map[types.UID]int32{}
 
+	// Look up the slice of names of pods for which we have already sent
+	// an event reporting either no controller or non-scale controller.
+	key, err := controller.KeyFunc(pdb)
+	if err != nil {
+		return []*v1.Pod{}, 0, err
+	}
+	knownNonScalePods, found := dc.eventCache[key]
+	if !found {
+		knownNonScalePods = make([]string, 0)
+	}
+
+	// nonScalePods will be populated with the names of pods for which
+	// scale can't be determined.
+	var nonScalePods []string
+
 	// 1. Find the controller for each pod.  If any pod has 0 controllers,
 	// that's an error. With ControllerRef, a pod can only have 1 controller.
 	for _, pod := range pods {
 		controllerRef := metav1.GetControllerOf(pod)
 		if controllerRef == nil {
-			err = fmt.Errorf("found no controller ref for pod %q", pod.Name)
-			dc.recorder.Event(pdb, v1.EventTypeWarning, "NoControllerRef", err.Error())
-			return
+			// Only create event if one hasn't been sent already.
+			if !slice.ContainsString(knownNonScalePods, pod.Name, nil) {
```

Comment:
The problem with this approach is that for controllers re-creating pods this might grow indefinitely and you're not cleaning that list anywhere other than when a PDB is removed. In the case described in #77383 the cronjob will end up creating jobs let's say every minute. You can easily fill that cache with names of those pods within a few minutes if the cronjob defines sufficiently big task, with hundreds of pods.

Reply:
So the list of pods are "recreated" on every reconcile, so the list will never contain more entries than the number of pods in the cluster that is covered by the PDB. If we don't want to keep this list, then it seems like we either need to accept that we will generate events for all non-conforming pods on every reconcile (which has scalability concerns), or change the way we handle events for PDBs to no longer send an event per pod, but instead just create one event per PDB (in which case maybe sending an event per reconcile is also less of an issue).